ext/erb/erb.c

```diff
@@ -0,0 +1,89 @@
+#include "ruby.h"
+#include "ruby/encoding.h"
+
+static VALUE rb_cERB, rb_mEscape;
+
+#define HTML_ESCAPE_MAX_LEN 6
+
+static const struct {
+    uint8_t len;
+    char str[HTML_ESCAPE_MAX_LEN+1];
+} html_escape_table[UCHAR_MAX+1] = {
+#define HTML_ESCAPE(c, str) [c] = {rb_strlen_lit(str), str}
+    HTML_ESCAPE('\'', "&#39;"),
+    HTML_ESCAPE('&', "&amp;"),
+    HTML_ESCAPE('"', "&quot;"),
+    HTML_ESCAPE('<', "&lt;"),
+    HTML_ESCAPE('>', "&gt;"),
+#undef HTML_ESCAPE
+};
+
+static inline void
+preserve_original_state(VALUE orig, VALUE dest)
+{
+    rb_enc_associate(dest, rb_enc_get(orig));
+}
+
+static inline long
+escaped_length(VALUE str)
+{
+    const long len = RSTRING_LEN(str);
+    if (len >= LONG_MAX / HTML_ESCAPE_MAX_LEN) {
+        ruby_malloc_size_overflow(len, HTML_ESCAPE_MAX_LEN);
+    }
+    return len * HTML_ESCAPE_MAX_LEN;
+}
+
+static VALUE
+optimized_escape_html(VALUE str)
+{
+    VALUE vbuf;
+    char *buf = ALLOCV_N(char, vbuf, escaped_length(str));
+    const char *cstr = RSTRING_PTR(str);
+    const char *end = cstr + RSTRING_LEN(str);
+
+    char *dest = buf;
+    while (cstr < end) {
+        const unsigned char c = *cstr++;
+        uint8_t len = html_escape_table[c].len;
+        if (len) {
+            memcpy(dest, html_escape_table[c].str, len);
+            dest += len;
+        }
+        else {
+            *dest++ = c;
+        }
+    }
+
+    VALUE escaped = str;
```

Comment:
As far as supporting ruby 2.7, C99 syntax cannot be used.

Reply:
What do you mean? Ruby 2.7 was the version when we started using C99 https://bugs.ruby-lang.org/projects/ruby-master/wiki/C99. Also, while it's a separate problem, ERB actually shouldn't support 2.7 because it became a default gem in Ruby 3.0.

Reply:
By the way, I believe https://github.com/ruby/cgi/blob/c1ffa3a428c706ef5dcef48ca057a76f3b839d75/ext/cgi/escape/escape.c#L66 was C99 as well.

Reply:
```c
$ cat a.c
#include <stdio.h>

int main(int argc, char **argv) {
    printf("code\n");
    int declaration;
    declaration = 1;
    printf("%d\n", declaration);
    return 0;
}

$ gcc -std=c89 -pedantic a.c
a.c: In function ‘main’:
a.c:5:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
    5 |     int declaration;
      |     ^~~
```

Reply:
Sorry, my fault.